src/backend/distributed/commands/function.c

```diff
@@ -91,6 +187,309 @@ CreateFunctionDDLCommandsIdempotent(const ObjectAddress *functionAddress)
 }
 
 
+/*
+ * GetDistributionArgIndex calculates the distribution argument with the given
+ * parameters. The function errors out if no valid argument is found.
+ */
+static int
+GetDistributionArgIndex(Oid functionOid, char *distributionArgumentName,
+						Oid *distributionArgumentOid)
+{
+	int distributionArgumentIndex = -1;
+
+	int numberOfArgs = 0;
+	int argIndex = 0;
+	Oid *argTypes = NULL;
+	char **argNames = NULL;
+	char *argModes = NULL;
+
+	HeapTuple proctup = NULL;
+
+	*distributionArgumentOid = InvalidOid;
+
+	proctup = SearchSysCache1(PROCOID, ObjectIdGetDatum(functionOid));
+	if (!HeapTupleIsValid(proctup))
+	{
+		elog(ERROR, "cache lookup failed for function %u", functionOid);
+	}
+
+	numberOfArgs = get_func_arg_info(proctup, &argTypes, &argNames, &argModes);
+
```

Comment:
Can we move `ReleaseSysCache(proctup);` up here?

Reply:
I think we probably should not  because we're  still using  `argTypes ` and `argNames ` below. Other callers of `get_func_arg_info` does keep until they're done processing with the `argTypes ` and `argNames `. Let's follow the postgres convention unless you have further objection?

Reply:
`get_func_arg_info` explicitly takes the trouble of allocating new memory and fresh new datums for these arrays, so they aren't tied to the tuple.

https://github.com/postgres/postgres/blob/REL_11_STABLE/src/backend/utils/fmgr/funcapi.c#L909